src/common/ad_tx.c

```diff
@@ -608,3 +612,397 @@ ad_tx_redo_act_next(struct ad_tx *tx)
 	}
 	return &tx->tx_redo_act_pos->it_act;
 }
+
+static struct ad_tx *
+tx_get()
+{
+	return tls_tx;
+}
+
+static void
+tx_set(struct ad_tx *tx)
+{
+	tls_tx = tx;
+}
+
+static void
+tx_callback(struct ad_tx *tx)
+{
+	if (!tx->tx_stage_cb || tx->tx_layer != 0)
+		return;
+	tx->tx_stage_cb(ad_tx_stage(tx), tx->tx_stage_cb_arg);
+}
+
+static int
+tx_end(struct ad_tx *tx, int err)
+{
+	struct umem_tx	*utx;
+	int		 rc;
+
+	if (err == 0)
+		err = tx->tx_last_errno;
+	else
+		tx->tx_last_errno = err;
+
+	D_ASSERTF(tx->tx_layer >= 0, "TX "DF_U64", bad layer %d\n", ad_tx_id(tx), tx->tx_layer);
+	if (tx->tx_layer != 0)
+		return 0;
+
+	rc = ad_tx_end(tx, err);
+	if (err == 0) {
+		if (rc == 0) {
```

Comment:
I think we should make sure "rc" is set to error code if @err is passed in, so we don't have to check twice?

Reply:
I think Liang's comment said checking 'rc' only should be fine. If err != 0, 'err' should be assigned to 'rc'.

Reply:
ok, i see. slight different with current PR's code, I'm fine to change it if need refresh this PR or later.